### PR TITLE
Cleanup HydroDyn and SeaState code

### PR DESCRIPTION
**Pull request by bjonkman**
**Feature or improvement description**
This PR removes duplicate copies of data and pointers of many variables shared between SeaState and HydroDyn. These shared data are now stored in the `WaveField` data type (stored as a SeaState parameter).

It also removes duplicated error checks, outdated code (including outdated driver codes for some HydroDyn submodules), and outdated comments. Hopefully these changes will make it easier to read the code, though more updates could be done. 

Some particular changes to note: 
- Parameter `WtrDpth` has been replaced with `EffWtrDpth` in places where it had added `MSL2SWL` to the variable. `WtrDpth` is now *ALWAYS* the value from the input file. This should help remove confusion about whether that variable is being used before or after adding  `MSL2SWL`.
- Several subroutines were modified so that SeaSt_Interp doesn't give the same warning about clamped boundaries hundreds of times when that error condition occurs.

**Impacted areas of the software**
HydroDyn and SeaState; glue codes that connect the two modules are also modified.

**Test results, if applicable**
Regression test results have not changed. 

Someone should add a test where `MSL2SWL` is non-zero, though. We should also add tests for the new features that have been added recently.

### CODE REVIEW

**Comment by andrew-platt** [comment]
@luwang00, can you also review?

File: modules/hydrodyn/src/Morison.f90 [hunk 1]
```diff
@@ -2126,7 +2117,7 @@ SUBROUTINE Morison_Init( InitInp, u, p, x, xd, z, OtherState, y, m, Interval, In
       tMG       = -999.0
       An_drag   = 0.0
       
-      IF ( (InitInp%InpJoints(i)%Position(3)-p%MSL2SWL) >= -p%WtrDpth ) THEN
+      IF ( (InitInp%InpJoints(i)%Position(3)-p%WaveField%MSL2SWL) >= -InitInp%WaveField%EffWtrDpth ) THEN !bjj: ask Lu if this is correct. I wonder if this check is supposed to be against WtrDpth
```
Comment (luwang00) [inline modules/hydrodyn/src/Morison.f90:2120]: This should be ok. Based on my notes, `InitInp%InpJoints(i)%Position(3)` is measured from the MSL. Alternatively, we can simplify the check to `InitInp%InpJoints(i)%Position(3) >= -InitInp%WaveField%WtrDpth`.

File: modules/seastate/src/SeaState_Input.f90 [hunk 2]
```diff
@@ -121,7 +122,7 @@ subroutine SeaSt_ParseInput( InputFileName, OutRootName, defWtrDens, defWtrDpth,
       if (Failed())  return;
 
       ! Z_Depth - Depth of the domain the Z direction.
-   call ParseVarWDefault ( FileInfo_In, CurLine, 'Z_Depth', InputFileData%Z_Depth, defWtrDpth+InputFileData%MSL2SWL, ErrStat2, ErrMsg2, UnEc )
+   call ParseVarWDefault ( FileInfo_In, CurLine, 'Z_Depth', InputFileData%Z_Depth, defWtrDpth+InputFileData%MSL2SWL, ErrStat2, ErrMsg2, UnEc ) !bjj: wouldn't the default be better with InputFileData%WtrDpth + InputFileData%MSL2SWL since we may have specified a WtrDpth already?
```
Comment (luwang00) [inline modules/seastate/src/SeaState_Input.f90:125]: @bjonkman I think you are right. We should change the default `Z_Depth` to `InputFileData%WtrDpth + InputFileData%MSL2SWL`.

File: modules/seastate/src/SeaState_Input.f90 [hunk 3]
```diff
@@ -743,8 +661,8 @@ subroutine SeaStateInput_ProcessInitData( InitInp, p, InputFileData, ErrStat, Er
 
 
        ! WaveHs - Significant wave height
-
-   if ( ( InputFileData%Waves%WaveMod /= 0 ) .AND. ( InputFileData%Waves%WaveMod /= 4 ) .AND. ( InputFileData%Waves%WaveMod /= 5 ) ) then   ! .TRUE. (when WaveMod = 1, 2, 3, or 10) if we have plane progressive (regular), JONSWAP/Pierson-Moskowitz spectrum (irregular) waves, or white-noise waves, but not user-defined or GH Bladed wave data.
+       !bjj: is this check still appropriate? do we need to add something for WaveMod 6 or 7? Otherwise, fix the comment on the next line
+   if ( ( InputFileData%WaveMod /= WaveMod_None ) .AND. ( InputFileData%WaveMod /= WaveMod_UserSpctrm ) .AND. ( InputFileData%WaveMod /= WaveMod_ExtElev ) ) then   ! .TRUE. (when WaveMod = 1, 2, 3, or 10) if we have plane progressive (regular), JONSWAP/Pierson-Moskowitz spectrum (irregular) waves, or white-noise waves, but not user-defined or GH Bladed wave data.
```
Comment (luwang00) [inline modules/seastate/src/SeaState_Input.f90:665]: @bjonkman Yes, I think we should add the conditions `InputFileData%WaveMod /= WaveMod_ExtFull` and `InputFileData%WaveMod /= WaveMod_UserFreq` to the `if` statement; however, it would be easier to read if we just switch the `if` and `else` branches.

File: modules/seastate/src/SeaState_Input.f90 [hunk 4]
```diff
@@ -759,122 +677,93 @@ subroutine SeaStateInput_ProcessInitData( InitInp, p, InputFileData, ErrStat, Er
...
-         call SetErrStat( ErrID_Info,'WvLowCOff adjusted to '//trim(num2lstr(TmpFreq))//' rad/s, based on WaveDT.',ErrStat,ErrMsg,RoutineName)
+      if ( InputFileData%WvHiCOff > TmpFreq ) then
+         InputFileData%WvHiCOff =  TmpFreq
+         call SetErrStat( ErrID_Info,'WvHiCOff adjusted to '//trim(num2lstr(TmpFreq))//' rad/s, based on WaveDT.',ErrStat,ErrMsg,RoutineName)
       end if
    end if
 
-   if (InputFileData%Waves%WaveMod > 2 .and. InputFileData%Waves%WaveMod /= 6) then
-      if ( InputFileData%Waves%WvLowCOff >= InputFileData%Waves%WvHiCOff ) then
+   !bjj: do we even need this check on WaveMod? Even if it's not being used, we can check that low < Hi, right?
```
Comment (luwang00) [inline modules/seastate/src/SeaState_Input.f90:712]: @bjonkman Agreed. There's no harm in requiring `WvHiCOff > WvLowCOff` even if they are not used. We can remove the `WaveMod` checks here.

File: modules/hydrodyn/src/HydroDyn_Output.f90 [hunk 5]
```diff
@@ -1006,9 +1006,10 @@ SUBROUTINE HDOUT_Init( HydroDyn_ProgDesc, OutRootName, InputFileData, y,  p, m,
    ! Check that the variables in OutList are valid      
    !-------------------------------------------------------------------------------------------------      
       
-   
-   CALL SetOutParam(InputFileData%OutList, p, ErrStat, ErrMsg )
-   IF ( ErrStat >= AbortErrLev ) RETURN
+   if (allocated(InputFileData%OutList)) then
```
Comment (luwang00) [inline modules/hydrodyn/src/HydroDyn_Output.f90:1009]: Not sure if this is related to Issue #1863 of Version 3.5.1.

File: modules/seastate/src/SeaState_Input.f90 (hunk 4 above)
Comment (ebranlard) [reply modules/seastate/src/SeaState_Input.f90:712]: As a user I prefer if the sanity checks are only done for the options that are activated. I like to put garbage in the inputs I don't use, to make it visually clear that I'm not using them.

Comment (luwang00) [reply modules/seastate/src/SeaState_Input.f90:712]: If we are to only perform the checks for the relevant `WaveMod` options, I think the ones that should be included are 2 (`WaveMod_JONSWAP`), 3 (`WaveMod_WhiteNoise`), 4 (`WaveMod_UserSpctrm`), 5 (`WaveMod_ExtElev`), and 7 (`WaveMod_UserFreq`) if I'm not mistaken. Not sure why 2 (JONSWAP) was not included in the original check. We might also want to overwrite `WvLowCOff` and `WvHiCOff` for the two regular-wave options in the code here just to make sure we are not cutting out that one wave component.

File: modules/hydrodyn/src/HydroDyn_Output.f90 (hunk 5 above)
Comment (deslaughter) [reply modules/hydrodyn/src/HydroDyn_Output.f90:1009]: This is a good check to perform, just in case OutList hasn't been allocated, but it's not causing the referenced issue.

File: modules/hydrodyn/src/WAMIT.f90 [hunk 6]
```diff
@@ -902,13 +901,14 @@ SUBROUTINE WAMIT_Init( InitInp, u, p, x, xd, z, OtherState, y, m, Interval, ErrS
       else
            ! Initialize the variables associated with the incident wave:
 
-         SELECT CASE ( InitInp%WaveMod ) ! Which incident wave kinematics model are we using?
-         CASE ( 0 )  ! No waves, NOTE: for this case we are forcing ExctnDisp = 0, so only p%WaveExctn needs to be allocated, not p%WaveExctnGrid
+         SELECT CASE ( InitInp%WaveField%WaveMod ) ! Which incident wave kinematics model are we using?
```
Comment (deslaughter) [inline modules/hydrodyn/src/WAMIT.f90:904]: Is there a reason that `InitInp%WaveField` isn't `p%WaveField`? It looks like it was changed in most other places.

**Review by deslaughter — APPROVED** [review]

File: modules/hydrodyn/src/HydroDyn_Input.f90 [hunk 7]
```diff
@@ -1191,42 +1165,24 @@ SUBROUTINE HydroDynInput_ProcessInitData( InitInp, Interval, InputFileData, ErrS
...
+         IF ( ( InitInp%WaveField%WaveStMod /= 0 ) .AND. ( InitInp%WaveField%WaveStMod /= 1 ) .AND. &
+              ( InitInp%WaveField%WaveStMod /= 2 ) .AND. ( InitInp%WaveField%WaveStMod /= 3 ) ) THEN
             ErrMsg  = ' WaveStMod must be 0, 1, 2, or 3.'
             ErrStat = ErrID_Fatal
             RETURN
          END IF
       ELSE
-         IF ( ( InitInp%WaveStMod /= 0 ) .AND. ( InitInp%WaveStMod /= 1 ) .AND. &
-              ( InitInp%WaveStMod /= 3 ) ) THEN
+         IF ( ( InitInp%WaveField%WaveStMod /= 0 ) .AND. ( InitInp%WaveField%WaveStMod /= 1 ) .AND. &
```
Comment (bjonkman) [inline modules/hydrodyn/src/HydroDyn_Input.f90:1177]: This check is different than the one in SeaSt_Input.f90/SeaStateInput_ProcessInitData(). In SeaState, it says WaveStMod must be 0 when WaveMod is WaveMod_ExtFull (6). Here in HydroDyn, it says it can be 0, 1, or 3. @luwang00 , do you know what the check should actually be?

File: modules/seastate/src/SeaState_Input.f90 [hunk 8]
```diff
@@ -589,129 +580,56 @@ subroutine SeaStateInput_ProcessInitData( InitInp, p, InputFileData, ErrStat, Er
...
+   IF ( InputFileData%WaveMod /= WaveMod_None .AND. InputFileData%WaveMod /= WaveMod_ExtFull ) THEN
+      IF ( (InputFileData%WaveStMod /= 0) .AND. (InputFileData%WaveStMod /= 1) .AND. &
+           (InputFileData%WaveStMod /= 2) .AND. (InputFileData%WaveStMod /= 3) ) THEN
          CALL SetErrStat( ErrID_Fatal,'WaveStMod must be 0, 1, 2, or 3.',ErrStat,ErrMsg,RoutineName)
          RETURN
       END IF
-   ELSE ! Wave stretching is not supported when WaveMod = 0 or 6.
-      InputFileData%Waves%WaveStMod = 0_IntKi
+   ELSE ! Wave stretching is not supported when WaveMod = 0 (WaveMod_None) or 6 (WaveMod_ExtFull).
+      InputFileData%WaveStMod = 0_IntKi
```
Comment (bjonkman) [inline modules/seastate/src/SeaState_Input.f90:614]: This is not consistent with the check in HydroDyn, where it says WaveStMod can be 0, 1, or 3 for WaveMod_ExtFull.

File: modules/hydrodyn/src/HydroDyn_Input.f90 (hunk 7 above)
Comment (luwang00) [reply modules/hydrodyn/src/HydroDyn_Input.f90:1177]: @bjonkman I believe the check performed in HydroDyn here is correct. Only extrapolation stretching (WaveStMod=2) cannot be used with full-field data. All other wave-stretching methods should work.

File: modules/seastate/src/SeaState_Input.f90 (hunk 8 above)
Comment (luwang00) [reply modules/seastate/src/SeaState_Input.f90:614]: @bjonkman This check in SeaState should be corrected following the check in HydroDyn above. WaveStMod=0, 1, and 3 should be allowed with WaveMod_ExtFull.

File: modules/hydrodyn/src/Morison.f90 (hunk 1 above)
Comment (bjonkman) [reply modules/hydrodyn/src/Morison.f90]: Addressed in 67c77c: I updated the check to be `InitInp%InpJoints(i)%Position(3) >= -InitInp%WaveField%WtrDpth`. I think that's easier to read.

File: modules/seastate/src/SeaState_Input.f90 (hunk 2 above)
Comment (bjonkman) [reply modules/seastate/src/SeaState_Input.f90]: Addressed in 67c77c: updated default value for `Z_Depth`

File: modules/seastate/src/SeaState_Input.f90 (hunk 3 above)
Comment (bjonkman) [reply modules/seastate/src/SeaState_Input.f90]: Addressed in 67c77c

File: modules/seastate/src/SeaState_Input.f90 (hunk 4 above)
Comment (bjonkman) [reply modules/seastate/src/SeaState_Input.f90]: Addressed in 67c77c: The check has been updated for relevant WaveMod options. For WaveMod options that are supposed to ignore these values, I set them to -HUGE() and +HUGE() so that when they ARE used in HydroDyn, they allow all frequencies.

File: modules/hydrodyn/src/HydroDyn_Output.f90 (hunk 5 above)
Comment (bjonkman) [reply modules/hydrodyn/src/HydroDyn_Output.f90:1009]: This was mostly to fix issues in Debug mode where it would error due to passing an invalid array reference.

File: modules/hydrodyn/src/WAMIT.f90 (hunk 6 above)
Comment (bjonkman) [reply modules/hydrodyn/src/WAMIT.f90]: Addressed in 67c77c: it was just an oversight. I think I had added the parameter p%WaveField after switching a few of the variables.

File: modules/hydrodyn/src/HydroDyn_Input.f90 (hunk 7 above)
Comment (bjonkman) [reply modules/hydrodyn/src/HydroDyn_Input.f90]: Addressed in 67c77c: This check was moved to SeaState, to replace the conflicting check there (and removed from HydroDyn)

File: modules/seastate/src/SeaState_Input.f90 (hunk 8 above)
Comment (bjonkman) [reply modules/seastate/src/SeaState_Input.f90]: Addressed in 67c77c: updated with the check from HydroDyn
